ko resolve segfaults when push=false

When push=false there are no publishers and multiPublisher.Publish will return a nil name.Reference causing a segfault:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x1610af5]

goroutine 13 [running]:
github.com/google/ko/pkg/resolve.ImageReferences.func1(0x1, 0x665a99)
        /usr/local/google/home/ianmi/go/pkg/mod/github.com/google/ko@v0.6.0/pkg/resolve/resolve.go:66 +0x115
golang.org/x/sync/errgroup.(*Group).Go.func1(0xc000362a80, 0xc0000b6230)
        /usr/local/google/home/ianmi/go/pkg/mod/golang.org/x/sync@v0.0.0-20200625203802-6e8e738ad208/errgroup/errgroup.go:57 +0x59
created by golang.org/x/sync/errgroup.(*Group).Go
        /usr/local/google/home/ianmi/go/pkg/mod/golang.org/x/sync@v0.0.0-20200625203802-6e8e738ad208/errgroup/errgroup.go:54 +0x66

--- a/pkg/commands/resolver.go
+++ b/pkg/commands/resolver.go
@@ -160,6 +160,16 @@ func makePublisher(po *options.PublishOptions) (publish.Interface, error) {
 			}
 			publishers = append(publishers, dp)
 		}
+
+		// If not publishing, at least generate a digest to simulate
+		// publishing.
+		if len(publishers) == 0 {
+			publishers = append(publishers, nopPublisher{
+				repoName: repoName,
+				namer:    namer,
+			})
+		}
+
 		return publish.MultiPublisher(publishers...), nil
 	}()
 	if err != nil {
@@ -169,6 +179,23 @@ func makePublisher(po *options.PublishOptions) (publish.Interface, error) {
 	// Wrap publisher in a memoizing publisher implementation.
 	return publish.NewCaching(innerPublisher)
 }
+
+// nopPublisher simulates publishing without actually publishing anything, to
+// provide fallback behavior when the user configures no push destinations.
+type nopPublisher struct {
+	repoName string
+	namer    publish.Namer
+}
+
+func (n nopPublisher) Publish(br build.Result, s string) (name.Reference, error) {
+	h, err := br.Digest()
+	if err != nil {
+		return nil, err
+	}
+	return name.NewDigest(fmt.Sprintf("%s/%s@%s", n.repoName, n.namer(s), h))
+}
+
+func (n nopPublisher) Close() error { return nil }
 
 // resolvedFuture represents a "future" for the bytes of a resolved file.
 type resolvedFuture chan []byte

--- a/pkg/publish/multi.go
+++ b/pkg/publish/multi.go
@@ -15,6 +15,8 @@
 package publish
 
 import (
+	"errors"
+
 	"github.com/google/go-containerregistry/pkg/name"
 	"github.com/google/ko/pkg/build"
 )
@@ -34,13 +36,16 @@ type multiPublisher struct {
 
 // Publish implements publish.Interface.
 func (p *multiPublisher) Publish(br build.Result, s string) (ref name.Reference, err error) {
+	if len(p.publishers) == 0 {
+		return nil, errors.New("MultiPublisher configured with zero publishers")
+	}
+
 	for _, pub := range p.publishers {
 		ref, err = pub.Publish(br, s)
 		if err != nil {
 			return
 		}
 	}
-
 	return
 }
 

--- a/pkg/publish/multi_test.go
+++ b/pkg/publish/multi_test.go
@@ -61,3 +61,15 @@ func TestMulti(t *testing.T) {
 		t.Errorf("Close() = %v", err)
 	}
 }
+
+func TestMulti_Zero(t *testing.T) {
+	img, err := random.Image(1024, 1)
+	if err != nil {
+		t.Fatalf("random.Image() = %v", err)
+	}
+
+	p := MultiPublisher() // No publishers.
+	if _, err := p.Publish(img, "foo"); err == nil {
+		t.Errorf("Publish() got nil error")
+	}
+}

